pkg/controllers/deprovisioning/consolidation.go

```diff
@@ -0,0 +1,286 @@
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	"github.com/aws/karpenter-core/pkg/scheduling"
+	"github.com/aws/karpenter-core/pkg/utils/pod"
+)
+
+// Consolidation is the consolidation controller.
+type Consolidation struct {
+	kubeClient    client.Client
+	cluster       *state.Cluster
+	provisioner   *provisioning.Provisioner
+	recorder      events.Recorder
+	clock         clock.Clock
+	cloudProvider cloudprovider.CloudProvider
+}
+
+func NewConsolidation(clk clock.Clock, kubeClient client.Client, provisioner *provisioning.Provisioner,
+	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster) *Consolidation {
+	return &Consolidation{
+		clock:         clk,
+		kubeClient:    kubeClient,
+		cluster:       cluster,
+		provisioner:   provisioner,
+		recorder:      recorder,
+		cloudProvider: cp,
+	}
+}
+
+// shouldNotBeDeprovisioned is a predicate used to filter deprovisionable nodes
+func (c *Consolidation) shouldNotBeDeprovisioned(_ context.Context, n *state.Node, provisioner *v1alpha5.Provisioner, _ []*v1.Pod) bool {
+	return provisioner == nil || provisioner.Spec.Consolidation == nil || !ptr.BoolValue(provisioner.Spec.Consolidation.Enabled) ||
+		n.Node.Annotations[v1alpha5.DoNotConsolidateNodeAnnotationKey] == "true"
+}
+
+// sortCandidates orders deprovisionable nodes by the disruptionCost
+func (c *Consolidation) sortCandidates(nodes []candidateNode) []candidateNode {
+	sort.Slice(nodes, func(i int, j int) bool {
+		return nodes[i].disruptionCost < nodes[j].disruptionCost
+	})
+	return nodes
+}
+
+// computeCommand generates a deprovisioning command given deprovisionable nodes
+//
+//nolint:gocyclo
+func (c *Consolidation) computeCommand(ctx context.Context, attempt int, candidates ...candidateNode) (deprovisioningCommand, error) {
+	// First delete any empty nodes we see
+	if attempt == 0 {
+		emptyNodes := lo.Filter(candidates, func(n candidateNode, _ int) bool { return len(n.pods) == 0 })
+		if len(emptyNodes) != 0 {
+			return deprovisioningCommand{
+				nodesToRemove: lo.Map(emptyNodes, func(n candidateNode, _ int) *v1.Node { return n.Node }),
+				action:        actionDeleteEmpty,
+				created:       c.clock.Now(),
+			}, nil
+		}
+	}
+
+	// If we get here, return nothingToDo as no candidateNodes are valid.
+	if attempt >= len(candidates) {
+		return deprovisioningCommand{action: actionDoNothing}, nil
+	}
+
+	node := candidates[attempt]
+
+	pdbs, err := NewPDBLimits(ctx, c.kubeClient)
+	if err != nil {
+		return deprovisioningCommand{action: actionFailed}, fmt.Errorf("tracking PodDisruptionBudgets, %w", err)
+	}
+	// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
+	// scheduling unless its really needed
+	if !canBeTerminated(node, pdbs) {
+		return deprovisioningCommand{action: actionNotPossible}, nil
+	}
+
+	defer metrics.Measure(deprovisioningDurationHistogram.WithLabelValues("Replace/Delete"))()
```

Comment:
Is this metric measuring what it should now?

Reply:
Yep, it's at the same point in time of the code.